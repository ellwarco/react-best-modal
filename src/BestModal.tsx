import * as React from 'react';
import Portal from './Portal';
import FocusLock from 'react-focus-lock';

export interface Props extends React.AllHTMLAttributes<any> {
  children: React.ReactNode;
  onRequestClose: (e: KeyboardEvent) => void;
  appRoot: Element;
  disableFocusLock?: boolean;
  // Allow through any properties that weren't picked up
  [x: string]: any;
}

const ESC_KEY = 27;

export default class BestModal extends React.Component<Props> {
  previousFocusedElement: HTMLElement;

  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown);
    this.props.appRoot.setAttribute('aria-hidden', 'true');
    this.previousFocusedElement = document.activeElement as HTMLElement;
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown);
    this.props.appRoot.removeAttribute('aria-hidden');

    // Something inside <FocusLock /> is setting activeElement.
    // This overrides that behaviour.
    setTimeout(() => this.previousFocusedElement.focus(), 0);
  }

  onKeyDown = (e: KeyboardEvent) => {
    if (e.keyCode === ESC_KEY) {
      this.props.onRequestClose(e);
    }
  };

  render() {
    const { children, onRequestClose, appRoot, disableFocusLock, ...props } = this.props;

    return (
      <Portal>
        <div aria-modal="true" role="dialog" {...props}>
          <FocusLock autoFocus={!disableFocusLock}>{children}</FocusLock>
        </div>
      </Portal>
    );
  }
}
