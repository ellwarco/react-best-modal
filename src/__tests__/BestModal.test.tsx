import * as React from 'react';
import BestModal from '../';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as sinon from 'sinon';

const noop = () => {};

describe('<BestModal />', () => {
  it('should pass through all extra props', () => {
    const appRoot = document.createElement('div');
    const extraProps = {
      'aria-labelledby': 'my-h1',
      'aria-describedby': 'my-p',
      className: 'my-modal',
    };

    const wrapper = shallow(
      <BestModal appRoot={appRoot} onRequestClose={noop} {...extraProps}>
        hi
      </BestModal>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should request to close when escape has been pressed', () => {
    const appRoot = document.createElement('div');
    const callback = sinon.spy();
    const addEventListener = sinon.stub(document, 'addEventListener');
    const wrapper = shallow(
      <BestModal appRoot={appRoot} onRequestClose={callback}>
        hi
      </BestModal>
    );
    const trigger = addEventListener.firstCall.args[1];
    const e = {
      keyCode: 27,
    };

    trigger(e);

    addEventListener.restore();
    expect(callback.calledWith(e)).toBe(true);
  });

  it('should do nothing when other keys are pressed', () => {
    const appRoot = document.createElement('div');
    const callback = sinon.spy();
    const addEventListener = sinon.stub(document, 'addEventListener');
    const wrapper = shallow(
      <BestModal appRoot={appRoot} onRequestClose={callback}>
        hi
      </BestModal>
    );
    const trigger = addEventListener.firstCall.args[1];
    const e = {
      keyCode: 10,
    };

    trigger(e);

    addEventListener.restore();
    expect(callback.called).toBe(false);
  });

  it('should listen to keydown on mount', () => {
    const appRoot = document.createElement('div');
    const callback = sinon.spy();
    const addEventListener = sinon.stub(document, 'addEventListener');

    const wrapper = shallow(
      <BestModal appRoot={appRoot} onRequestClose={callback}>
        hi
      </BestModal>
    );

    addEventListener.restore();
    expect(addEventListener.firstCall.args[0]).toEqual('keydown');
  });

  it('should unlisten to keydown when unmounting', () => {
    const appRoot = document.createElement('div');
    const callback = sinon.spy();
    const addEventListener = sinon.stub(document, 'addEventListener');
    const removeEventListener = sinon.stub(document, 'removeEventListener');
    const wrapper = shallow(
      <BestModal appRoot={appRoot} onRequestClose={callback}>
        hi
      </BestModal>
    );

    wrapper.unmount();

    addEventListener.restore();
    removeEventListener.restore();
    expect(removeEventListener.firstCall.args[0]).toEqual('keydown');
    expect(removeEventListener.firstCall.args[1]).toEqual(addEventListener.firstCall.args[1]);
  });

  it('should set aria-hidden to appRoot when mounting', () => {
    const appRoot = document.createElement('div');

    const wrapper = shallow(
      <BestModal appRoot={appRoot} onRequestClose={noop}>
        hi
      </BestModal>
    );

    expect(appRoot.getAttribute('aria-hidden')).toBe('true');
  });

  it('should remove aria-hidden from appRoot when unmounting', () => {
    const appRoot = document.createElement('div');
    const wrapper = shallow(
      <BestModal appRoot={appRoot} onRequestClose={noop}>
        hi
      </BestModal>
    );

    wrapper.unmount();

    expect(appRoot.hasAttribute('aria-hidden')).toBe(false);
  });

  it('should restore focus to previous element when unmounting', () => {
    const previousElement = document.createElement('button');
    previousElement.focus();
    const appRoot = document.createElement('div');
    const wrapper = mount(
      <BestModal appRoot={appRoot} onRequestClose={noop}>
        <button>close</button>
        <input />
      </BestModal>
    );

    wrapper.unmount();

    expect(document.activeElement).toBe(previousElement);
  });

  it('should focus on first focusable element when mounting', done => {
    const appRoot = document.createElement('div');
    const wrapper = mount(
      <BestModal appRoot={appRoot} onRequestClose={noop}>
        <button data-its-me>close</button>
        <input />
      </BestModal>
    );

    setImmediate(() => {
      expect(document.activeElement.hasAttribute('data-its-me')).toBe(true);
      done();
    });
  });
});
