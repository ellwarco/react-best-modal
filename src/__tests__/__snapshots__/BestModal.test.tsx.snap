// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BestModal /> should pass through all extra props 1`] = `
<Portal>
  <FocusLock
    allowTextSelection={false}
    disabled={false}
    noFocusGuards={false}
    returnFocus={false}
    sandboxed={false}
  >
    <div
      aria-describedby="my-p"
      aria-labelledby="my-h1"
      aria-modal="true"
      className="my-modal"
      role="dialog"
    >
      hi
    </div>
  </FocusLock>
</Portal>
`;
